presto-spi/src/main/java/com/facebook/presto/spi/block/MapBlock.java

```diff
@@ -303,9 +296,48 @@ public Block getLoadedBlock()
                 offsets,
                 keyBlock,
                 loadedValueBlock,
-                hashTables,
+                Optional.ofNullable(hashTables),
                 keyType,
                 keyBlockNativeEquals,
-                keyNativeHashCode);
+                keyNativeHashCode,
+                keyBlockHashCode);
+    }
+
+    @Override
+    protected void ensureHashTableLoaded()
+    {
+        if (this.hashTables == null) {
+            // This can only happen for MapBlock, not MapBlockBuilder because the latter always has non-null hashtables
+            synchronized (this) {
+                int[] offsets = getOffsets();
+                int elementCount = getRawKeyBlock().getPositionCount();
+                int mapCount = getPositionCount();
+                boolean[] mapIsNull = getMapIsNull();
+
+                if (this.hashTables == null) {
```

Comment:
Also, invert this one, and move to right after the start of the synchronized block.